### PR TITLE
[v2] [cache-dir] Use 404 resources in EnsureResources if the page doesn't exist

### LINKED ISSUE #8297: Page 404 redirects to ?no-cache=1
## Description

I'm using Gatsby V2. I have a strange problem with 404 error because it redirects automatically to ?no-cache=1

My URL is: https://aikode-pwa-sistema.firebaseapp.com/

### Steps to reproduce

I executed:
- gatsby new gatsby-site https://github.com/gatsbyjs/gatsby-starter-default
- I added some pages
- gatsby build
- firebase deploy (or gatsby serve) 
- enter a page that does not exist, for example: 
https://aikode-pwa-sistema.firebaseapp.com/page-test-404

### Expected result

show URL without "?no-cache=1"

https://aikode-pwa-sistema.firebaseapp.com/page-test-404

### Actual result

show URL with "?no-cache=1"

https://aikode-pwa-sistema.firebaseapp.com/page-test-404?no-cache=1

### Environment

System:
    OS: Windows 10
    CPU: x64 Intel(R) Core(TM) i7-3632QM CPU @ 2.20GHz
  Binaries:
    Yarn: 1.7.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 5.6.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: 42.17134.1.0
  npmPackages:
    gatsby: ^2.0.0 => 2.0.1
    gatsby-plugin-manifest: ^2.0.2 => 2.0.2
    gatsby-plugin-nprogress: ^2.0.5 => 2.0.5
    gatsby-plugin-offline: ^2.0.5 => 2.0.5
    gatsby-plugin-react-helmet: ^3.0.0 => 3.0.0
    gatsby-plugin-remove-trailing-slashes: ^2.0.0 => 2.0.0

### PR DESCRIPTION
**Pull request by davidbailey00**
- Get the 404 resources in `ensure-resources.js`
- Separate code in `load-directly-or-404.js` to prevent code reuse
- Return 404 resources from `getResourcesFromPathnameSync` if the page isn't found, to match behaviour of the async version
- Minor refactors to improve performance based on the changes made

Fixes #8297

### CODE REVIEW

File: packages/gatsby/cache-dir/ensure-resources.js [hunk 1]
```diff
@@ -93,24 +95,19 @@ class EnsureResources extends React.Component {
...
-      const url = new URL(window.location)
-      if (url.search) {
-        url.search += `&no-cache=1`
-      } else {
-        url.search = `?no-cache=1`
-      }
-      window.location.replace(url)
-
+      const url = getRedirectUrl(this.state.location.pathname)
+      if (url) window.location.replace(url)
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/ensure-resources.js]: Could you use the non-shorthand form?

File: packages/gatsby/cache-dir/ensure-resources.js [hunk 2]
```diff
@@ -16,6 +17,7 @@ class EnsureResources extends React.Component {
     this.state = {
       location: { ...location },
```
Comment (DSchau) [inline packages/gatsby/cache-dir/ensure-resources.js]: Where did we ever end up on this? Is this needed?

Comment (pieh) [reply packages/gatsby/cache-dir/ensure-resources.js]: hah, probably not! just don't use `git blame` - it tells lies!

**Comment by pieh** [comment]
I chatted with @davidbailey00 a little and we (or just him) will iterate on this tomorrow

**Pull request merged**